### PR TITLE
Adding global caches to survive/warn on HMR, cleanup, etc.

### PR DESCRIPTION
**Pull request by jamesdaniels**
* Adding global instance caches to the modules, so they don't freak out when HMR is enabled (#2655). This takes on `globalThis` as a needed polyfill for many environments.
* If injected settings for modules are changed after they are initialized you will receive a warning and the prior instance will be returned (ignoring the changes), this is especially important for HMR. If HMR is detected there will be an additional warning suggesting they do a full reload to see the changes.
* Added a polyfill table and notes about why we version like we do
* Adding more convoluted stuff to my sample app to flex AngularFire
* Internal cleanup on AngularFireAnalytics
* AngularFireAnalytics will now wait for UserTrackingService to detect the user before sending the screen_view event, if UserTrackingService has been provided
* Adding a warning if the Auth Emulator is detected in conjunction with AngularFirestore and AngularFireDatabase as they will invalidate the emulated auth token before the dynamic import of `firebase/auth` is completed (#2656)
* Warn if we absorbed an error keeping Firestore persistence from enabling
* Logging sign_up and login events in UserTrackingService
* Adding credential observer to AngularFireAuth

### CODE REVIEW

File: src/analytics/user-tracking.service.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        this.disposable = auth.user.pipe(
+            tap(user => {
+                // TODO pairwise, null => !null track login
+                analytics.setUserId(user?.uid);
+                resolve();
+            }),
+            pairwise(),
+            tap(([prev, current]) => {
+                if (!prev && current) {
+                    // TODO how do I differentiate from signup? current.metadata.creationTime
```
Comment (jamesdaniels) [inline src/analytics/user-tracking.service.ts:33]: Remove. Can't do this the .user, need to watch the signIn* returns.

File: src/analytics/user-tracking.service.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    // tslint:disable-next-line:ban-types
+    @Inject(PLATFORM_ID) platformId: Object,
+    auth: AngularFireAuth,
+  ) {
+
+    this.initialized = new Promise(resolve => {
+      if (!isPlatformServer(platformId)) {
+        this.disposable = auth.user.pipe(
+            tap(user => {
+                // TODO pairwise, null => !null track login
```
Comment (jamesdaniels) [inline src/analytics/user-tracking.service.ts:26]: Remove todo

File: src/firestore/firestore.ts [hunk 3]
```diff
@@ -136,43 +145,44 @@ export class AngularFirestore {
...
+        // https://github.com/firebase/firebase-js-sdk/issues/608
+        const enablePersistence = () => {
+          try {
+            return from(firestore.enablePersistence(persistenceSettings || undefined).then(() => true, () => false));
+          } catch (e) {
+            if (typeof console !== 'undefined') { console.warn(e); }
+            return of(false);
+          }
+        };
+        // TODO how do I back this up for a cache hit?
```
Comment (jamesdaniels) [inline src/firestore/firestore.ts:179]: remove todo addressed.

**Comment by volago** [comment]
Hi.

I've notice that credentials are available just after user logging. When I refresh the page, user is still logged, but credentials are `null`. Is it desired behavior? 

Where can I get accessToken (for download Facebook avatar) when user is already logged in?

**Comment by jamesdaniels** [comment]
@volago Yes, this is expected behavior with Firebase. If you need persistence for the access token you'll need to do that yourself, toss it in Firestore/Database perhaps? Make sure to lock it down with security rules.

**Pull request merged**